[Clang][Basic] Add `__has_feature` checks for CFI sanitizers

moorabbit opened the pull request:
Add `__has_feature` checks control flow integrity sanitizers.

Fixes #151022

moorabbit committed:
[Clang][Basic] Enable `__has_feature(cfi)`

Support `__has_feature(cfi)` to check for control flow integrity sanitizers.

PiJoules reviewed (commented):
Thanks for taking this. Just a few comments from me.

moorabbit committed:
Add more test cases

PiJoules reviewed (approved):
LGTM thanks

moorabbit committed:
add `-fno-sanitize-ignorelist` to fix CI build issue

moorabbit commented:
My apologies for the force-push. Issue with git.

CI was failing with the error `missing sanitizer ignorelist`. My last commit adds the `-fno-sanitize-ignorelist` flag to some clang invocations in order to fix the issue.

PiJoules commented on clang/test/Lexer/has_feature_cfi.c:
```diff
@@ -0,0 +1,31 @@
+// RUN: %clang -E --target=x86_64-linux-gnu -fvisibility=hidden -fno-sanitize-ignorelist -fsanitize=cfi -fno-sanitize-cfi-cross-dso -fno-sanitize=cfi-nvcall,cfi-vcall,cfi-mfcall,cfi-icall -flto -c %s -o - | FileCheck %s --check-prefix=CHECK-CFISAN
+
+// Disable all CFI schemes. This essentially disables CFI sanitizers.
+// RUN: %clang -E --target=x86_64-linux-gnu -fvisibility=hidden -fsanitize=cfi -fno-sanitize-cfi-cross-dso -fno-sanitize=cfi-nvcall,cfi-vcall,cfi-mfcall,cfi-icall,cfi-cast-strict,cfi-derived-cast,cfi-unrelated-cast -flto -c %s -o - | FileCheck %s --check-prefix=CHECK-NO-CFISAN
```
If you're testing for x86_64-linux-gnu only, you'll probably want `// REQUIRES: x86-registered-target` also

moorabbit replied:
On second thought, since I'm just running the preprocessor (i.e. `clang -E`), I probably don't need to set the target at all. I removed the target specification in the new commit. Let me know if you would like me to keep it and require `x86-registered-target`.

moorabbit committed:
Fix driver test failure

moorabbit commented:
> -fno-sanitize-ignorelist shouldn't be needed here. What errors were you originally seeing?

Here's the complete error message I get locally (CI shows the same):
```
clang: error: missing sanitizer ignorelist: '[...]/build/lib/clang/22/share/cfi_ignorelist.txt'
```

moorabbit commented:
There's also a problem that caused a driver test to fail. I fixed it in [b6a7ce7](https://github.com/llvm/llvm-project/pull/151348/commits/b6a7ce7e3f4f51ca71f813d931362a710abd7d87). The issue was that the driver was not expecting `-fsanitize=cfi` to also enable the `cfi-cast-strict` scheme.

After doing a little bit of digging, I found out that the [original commit](https://github.com/llvm/llvm-project/commit/d2926c91d5135a95912fdc142d133b647668e918#diff-a7ab5c13e765c93093f11fe7283020fcdbe13dfbc0e18e80cf29cfe90d8e9b4bR82) that added the CFI cast sanitizers didn't enable the `cfi-cast-strict` scheme by default (i.e. with `-fsanitize=cfi`). You need to explicitly specify `-fsanitize=cfi-cast-strict` in order to have it enabled. So I removed, in the last commit, `CFICastStrict` from the CFI sanitizer group.

Ran `check-clang` locally and didn't find any issues. CI should pass now.

pcc reviewed (changes requested):
cc @fmayer 

We have been trying to break `__has_feature(undefined_behavior_sanitizer)` into fine grained checks (#148310). If we have `__has_feature(cfi)`mean CFI || KCFI that is sort of going in the opposite direction. Let's take some time to think about whether it is a good idea to combine them like this.

fmayer commented:
> Also, I just learned that non-standard features are meant to be under `__has_extension` (`EXTENSION` macro) instead of `__has_feature`. See
> 
> https://github.com/llvm/llvm-project/blob/154354e60e46debb68716cde867cb4e325ce5eb7/clang/include/clang/Basic/Features.def#L20

Given `kcfi` already is in `has_feature`, maybe (if we decide to add this) it still is preferable to also put `cfi` there?

PiJoules commented:
@moorabbit could you split the check into individual ones akin to https://github.com/llvm/llvm-project/pull/148310

moorabbit committed:
PiJoules suggestion

PiJoules reviewed (commented):
LGTM but let's also wait for approval from either @pcc or @fmayer

fmayer commented on clang/test/Lexer/has_feature_cfi.c:
```diff
@@ -0,0 +1,77 @@
+// RUN: %clang -E -fvisibility=hidden -flto -fno-sanitize-ignorelist -fsanitize=cfi -fsanitize-cfi-cross-dso -c %s -o - | FileCheck %s --check-prefix=CHECK-CFI
+// RUN: %clang -E -fvisibility=hidden -flto -fno-sanitize-ignorelist -fsanitize=cfi -fno-sanitize=cfi-nvcall,cfi-vcall,cfi-mfcall,cfi-icall -c %s -o - | FileCheck %s --check-prefix=CHECK-CFI
+// RUN: %clang -E -fsanitize=kcfi -c %s -o - | FileCheck %s --check-prefix=CHECK-CFI
+// CHECK-CFI: CFISanitizerEnabled
```
I agree to separate them, but I don't think we can change that later as soon as anyone uses that.

moorabbit replied:
Sounds good. Tried to fix it in a new commit.

moorabbit committed:
PiJoules suggestions

moorabbit commented on clang/include/clang/Basic/Features.def:
```diff
@@ -303,6 +303,14 @@ FEATURE(is_trivially_assignable, LangOpts.CPlusPlus)
 FEATURE(is_trivially_constructible, LangOpts.CPlusPlus)
 FEATURE(is_trivially_copyable, LangOpts.CPlusPlus)
 FEATURE(is_union, LangOpts.CPlusPlus)
+FEATURE(cfi_sanitizer, LangOpts.Sanitize.hasOneOf(SanitizerKind::CFI | SanitizerKind::CFICastStrict | SanitizerKind::KCFI))
```
Fixed in a new commit. Thanks!

fmayer reviewed (approved):
lgtm thanks

small nit: I would say Add or Support rather than Enable in the first line of the commit message

moorabbit committed:
Fix test failure on windows

moorabbit committed:
Specify the target platform the test is allowed to run on

moorabbit committed:
Merge branch 'main' into has-feature-cfi

fmayer commented:
> Added a REQUIRES: target={{x86_64.*-linux.*}} in my last commit to fix the test failure on windows (-fsanitize=cfi-mfcall is not supported on windows) .
Since this PR is only adding a preprocessor feature and not something platform specific, I think it's fine to only test on x86_64-linux and skip other targets?


Yeah, I did the same with the UBSan change.

moorabbit committed:
Merge branch 'main' into has-feature-cfi

fmayer commented:
Thanks!
